safehttp/task.go

```diff
@@ -0,0 +1,256 @@
+	t.writtenError = true
+	resp := &ErrorResponse{Code: code}
+	t.errorPhase(resp)
+	http.Error(t.rw, http.StatusText(int(resp.Code)), int(resp.Code))
+	return Result{}
+}
+
+// Redirect responds with a redirect to the given url, using code as the status code.
+//
+// If the ResponseWriter has already been written to, then this method will panic.
+func (t *Task) Redirect(r *IncomingRequest, url string, code StatusCode) Result {
+	if code < 300 || code >= 400 {
+		panic("wrong method called")
+	}
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	http.Redirect(t.rw, r.req, url, int(code))
+	return Result{}
+}
+
+// Header returns the collection of headers that will be set
+// on the response. Headers must be set before writing a
+// response (e.g. Write, WriteTemplate).
+func (t *Task) Header() Header {
+	return t.header
+}
+
+// SetCookie adds a Set-Cookie header to the provided ResponseWriter's headers.
+// The provided cookie must have a valid Name, otherwise an error will be
+// returned.
+func (t *Task) SetCookie(c *Cookie) error {
+	return t.header.addCookie(c)
+}
+
+// SetCode allows setting a response status. If the response was already
+// written, trying to set the status code will have no effect. This method will
+// panic if an invalid status code is passed (i.e. not in the range 1XX-5XX).
+//
+// If SetCode was called before NoContent, Redirect or WriteError, the status
+// code set by the latter will be the actual response status.
+//
+// TODO(empijei@, kele@, maramihali@): decide what should be done if the
+// code passed is either 3XX (redirect) or 4XX-5XX (client/server error).
+func (t *Task) SetCode(code StatusCode) {
+	if t.written {
+		return
+	}
+	if code < 100 || code >= 600 {
+		panic("invalid status code")
+	}
+	t.code = code
+}
+
+// commitPhase calls the Commit phases of all the interceptors. This stage will
+// run before a response is written to the ResponseWriter. If a response is
+// written to the ResponseWriter in a Commit phase then the Commit phases of the
+// remaining interceptors won't execute.
+//
+// TODO: BIG WARNING, if ResponseWriter.Write and ResponseWriter.WriteTemplate
+// are called in Commit then this will recurse. CommitResponseWriter was an
+// attempt to prevent this by not giving access to Write and WriteTemplate in
+// the Commit phase.
+func (t *Task) commitPhase(resp Response) {
+	for i := len(t.cfg.Interceptors) - 1; i >= 0; i-- {
+		t.cfg.Interceptors[i].Commit(t, t.req, resp)
+	}
+}
+
+// errorPhase calls the OnError phases of all the interceptors associated with
+// a handler. This stage runs before an error response is written to the
+// ResponseWriter.
+//
+// TODO: BIG WARNING, if an interceptor attempts to write to the ResponseWriter
+// in the OnError phase will result in an irrecoverable error.
+func (t *Task) errorPhase(resp Response) {
+	for i := len(t.cfg.Interceptors) - 1; i >= 0; i-- {
+		t.cfg.Interceptors[i].OnError(t, t.req, resp)
```

Comment:
Should we maybe pass a smaller interface to OnError that doesn't have Write* methods so we make sure Write is not called in this step?

Reply:
I'm not sure. It's not obvious to me what `OnError` should and shouldn't be able to do. How much should it be able to modify the response?

Reply:
sometimes we call it after the response has been partially written, there isn't much it can modify at that point, is there?

Reply:
The current code is broken, sure. But I'd like to somehow gracefully handle panics occurred in the handler. In this case, our framework should (maybe) call `OnError` of all the interceptors, in which case maybe they should be able to do writes?